keystone/gauss_fit.py

```diff
@@ -0,0 +1,230 @@
+	# with SNR above snr_min
+	x = []
+	y = []
+	pixels = 0
+	for (i,j), value in np.ndenumerate(cube_gauss[0]):
+     		spectra=np.array(cube_km.unmasked_data[:,i,j])
+     		rms = np.std(np.append(spectra[0:(peak_channels[0]-1)], spectra[(peak_channels[1]+1):len(spectra)]))
+     		if (max(spectra[peak_channels[0]:peak_channels[1]]) / rms) > snr_min:
+            		pixels+=1
+	    		x.append(i)
+	    		y.append(j)
+     		else:	
+	    		cube_gauss[:,i,j]=nan_array
+	    		param_cube[:,i,j]=nan_array2
+			cube_gauss_noise[:,i,j]=nan_array
+	print str(pixels) + ' Pixels above SNR=' + str(snr_min) 
+
+	# Define a Gaussian fitting function for each pixel
+	# i, j are the x,y coordinates of the pixel being fit
+	def pix_fit(i,j):
+		spectra = np.array(cube_km.unmasked_data[:,i,j])
+		# Use the peak brightness Temp within specified channel 
+		# range as the initial guess for Gaussian height
+		Tpeak = max(spectra[peak_channels[0]:peak_channels[1]])
+		# Use the velocity of the brightness Temp peak as 
+		# initial guess for Gaussian mean
+		vpeak = spectra_x_axis_kms[peak_channels[0]:peak_channels[1]][np.where(spectra[peak_channels[0]:peak_channels[1]]==Tpeak)]
+		rms = np.std(np.append(spectra[0:(peak_channels[0]-1)], spectra[(peak_channels[1]+1):len(spectra)]))
+		err1 = np.zeros(shape[0])+rms
+		# Create a noise spectrum based on rms of off-line channels
+		# This will be added to best-fit Gaussian to obtain a noisy Gaussian 
+		noise=np.random.uniform(-1.*rms,rms,len(spectra_x_axis_kms))
+		# Define initial guesses for Gaussian fit
+		guess = [Tpeak, vpeak, 0.3] # [height, mean, sigma]
+		try:
+			coeffs, covar_mat = curve_fit(p_eval, xdata=spectra_x_axis_kms, ydata=spectra, p0=guess, sigma=err1, maxfev=500)
+			gauss = np.array(p_eval(spectra_x_axis_kms,coeffs[0], coeffs[1], coeffs[2]))
+			noisy_gauss = np.array(p_eval(spectra_x_axis_kms,coeffs[0], coeffs[1], coeffs[2]))+noise
+			params = np.append(coeffs, (covar_mat[0][0]**0.5, covar_mat[1][1]**0.5, covar_mat[2][2]**0.5)) 
+			# params = ['Tpeak', 'VLSR','sigma','Tpeak_err','VLSR_err','sigma_err']
+
+			# Don't accept fit if fitted parameters are non-physical or too uncertain
+			if (params[0] < 0.01) or (params[3] > 1.0) or (params[2] < 0.05) or (params[5] > 0.5) or (params[4] > 0.75):
+				noisy_gauss = nan_array
+				gauss = nan_array
+				params = nan_array2
+
+			# Don't accept fit if the SNR for fitted spectrum is less than SNR threshold
+			#if max(gauss)/rms < snr_min:
+			#	noisy_gauss = nan_array
+			#	gauss = nan_array
+			#	params = nan_array2
+		
+		except RuntimeError:
+			noisy_gauss = nan_array
+			gauss = nan_array
+			params = nan_array2
+	
+		return i, j, gauss, params, noisy_gauss
+
+	# Parallel computation:
+	nproc = multicore  # maximum number of simultaneous processes desired
+	queue = pprocess.Queue(limit=nproc)
+	calc = queue.manage(pprocess.MakeParallel(pix_fit))
+	tic=time.time()
+	counter = 0
+
+	# Uncomment to see some plots of the fitted spectra
```

Comment:
Plots can be nice to have without reinstalling the code.  You could think about adding a `doPlot=False` keyword and wrap the block in an `if doPlot:` call to get it if desired.

Reply:
Yes, I should add that. Those plots were mainly for me during testing but it should still be an option in the function